### PR TITLE
tests: updateFeeAccumulatorPosition

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #3980 

## What is the purpose of the change

Adds test cases to fully cover updateFeeAccumulatorPosition

## Brief Changelog

The following vectors were tested:
 
- updating a position with a different pool id implies updating a different position
- updating a position with a different owner implies updating a different position
- updating a position with a  different lowe tick implies updating a different position
- updating a position with an upper tick implies updating a different position

This PR is directed at roman/create-position-fees since it is not yet merged and utilizes the updateFeeAccumulatorPosition. If it were directed at main, there would be merge conflicts if this PR or that PR was merged.

### CODE REVIEW

File: x/concentrated-liquidity/fees.go [hunk 1]
```diff
@@ -91,8 +91,9 @@ func (k Keeper) initializeFeeAccumulatorPosition(ctx sdk.Context, poolId uint64,
 	return nil
 }
 
-// updateFeeAccumulatorPosition updates the owner's position
-// TODO: test
+// updateFeeAccumulatorPosition updates the fee accumulator position for a given pool, owner, and tick range.
+// It retrieves the fee growth outside of the given tick range and updates the position's accumulator
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees.go:95]: nit:
```suggestion
// It retrieves the current fee growth outside of the given tick range and updates the position's accumulator
```

File: x/concentrated-liquidity/fees_test.go [hunk 2]
```diff
@@ -699,3 +699,114 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+		},
+	}
+
+	for name, tc := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+
+			// Setup two cl pools
+			poolOne := s.PrepareConcentratedPool()
+			poolTwo := s.PrepareConcentratedPool()
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees_test.go:770]: nit: would it make sense to have a shared state between test cases and prepare both pools up top?

File: x/concentrated-liquidity/fees_test.go [hunk 3]
```diff
@@ -699,3 +699,114 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+
+			pools := []cltypes.ConcentratedPoolExtension{poolOne, poolTwo}
+			owners := []sdk.AccAddress{ownerOne, ownerTwo}
+
+			// Initialize three base positions in each pool for each owner (total of 12 positions)
+			for _, pos := range positions {
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees_test.go:776]: We could also just initialize this once and for all the test cases

File: x/concentrated-liquidity/fees_test.go [hunk 4]
```diff
@@ -699,3 +699,114 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+			upperTick: DefaultUpperTick,
+		},
+		"update position with a different lower tick": {
+			poolId:    1,
+			owner:     ownerOne,
+			liquidity: DefaultLiquidityAmt,
+			lowerTick: DefaultLowerTick + 1,
+			upperTick: DefaultUpperTick,
+		},
+		"update position with a different upper tick": {
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees_test.go:755]: It should be possible to test error cases too. Can we cover cases such as pool does not exist, fee accumulator position is not pre-initialized.

I think the latter is very important to cover

File: x/concentrated-liquidity/fees_test.go [hunk 5]
```diff
@@ -699,3 +699,114 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+			// System under test
+			// Update one of the positions as per the test case
+			err := s.App.ConcentratedLiquidityKeeper.UpdateFeeAccumulatorPosition(s.Ctx, tc.poolId, tc.owner, tc.liquidity, tc.lowerTick, tc.upperTick)
+			s.Require().NoError(err)
+
+			if tc.expectedError != nil {
+				s.Require().Error(err)
+				s.Require().ErrorAs(err, &tc.expectedError)
+				return
+			}
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees_test.go:793]: nit: currently no error test case. As per other comment, we should have some

**Comment by p0mvn** [comment]
LGTM once requested tests added

**Comment by p0mvn** [comment]
@czarcas7ic merged main into your branch because my PR broke it

File: x/concentrated-liquidity/fees_test.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/fees_test.go:770]: I don't know if I agree with table driven test cases sharing state between runs. This is more so the job of the simulator.

File: x/concentrated-liquidity/fees_test.go (hunk 3 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/fees_test.go:776]: Same point as above

**Comment by czarcas7ic** [comment]
@p0mvn, I added two more test cases:
* pool does not exists
* position does not exits

File: x/concentrated-liquidity/fees_test.go [hunk 6]
```diff
@@ -699,3 +701,125 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+		},
+		"err: position does not exist": {
+			poolId:        1,
+			owner:         ownerOne,
+			liquidity:     DefaultLiquidityAmt,
+			lowerTick:     DefaultLowerTick - 1,
+			upperTick:     DefaultUpperTick,
+			expectedError: accum.NoPositionError{Name: fmt.Sprintf("%s,%s,%d,%d", "1", ownerOne.String(), DefaultLowerTick-1, DefaultUpperTick)},
+		},
+	}
```
Comment (mattverse) [inline x/concentrated-liquidity/fees_test.go:776]: wdyt of adding test cases for trying to updateFeeAccumulatorPosition for pool that does not fee accumulator?

File: x/concentrated-liquidity/fees_test.go [hunk 7]
```diff
@@ -699,3 +701,125 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+			expectedError: cltypes.PoolNotFoundError{PoolId: 3},
+		},
+		"err: position does not exist": {
+			poolId:        1,
+			owner:         ownerOne,
+			liquidity:     DefaultLiquidityAmt,
+			lowerTick:     DefaultLowerTick - 1,
+			upperTick:     DefaultUpperTick,
+			expectedError: accum.NoPositionError{Name: fmt.Sprintf("%s,%s,%d,%d", "1", ownerOne.String(), DefaultLowerTick-1, DefaultUpperTick)},
+		},
```
Comment (mattverse) [inline x/concentrated-liquidity/fees_test.go:775]: Does this go over both cases where this method works as both negative and positive? (`RemoveFromPositionCustomAcc`, `AddToPositionCustomAcc`) ?

**Review by mattverse — COMMENTED** [review]
LGTM once my minor comments / questions have been addressed!

File: x/concentrated-liquidity/fees_test.go [hunk 8]
```diff
@@ -699,3 +701,125 @@ func (s *KeeperTestSuite) TestCollectFees() {
...
+			poolId:        3,
+			expectedError: cltypes.PoolNotFoundError{PoolId: 3},
+		},
+		"err: position does not exist": {
+			poolId:        1,
+			owner:         ownerOne,
+			liquidity:     DefaultLiquidityAmt,
+			lowerTick:     DefaultLowerTick - 1,
+			upperTick:     DefaultUpperTick,
+			expectedError: accum.NoPositionError{Name: fmt.Sprintf("%s,%s,%d,%d", "1", ownerOne.String(), DefaultLowerTick-1, DefaultUpperTick)},
```
Comment (p0mvn) [inline x/concentrated-liquidity/fees_test.go:774]: nit: there is a helper function to perform this formatting: https://github.com/osmosis-labs/osmosis/blob/b2e2d3c5d7fe94ea9c3c87fe074dbc05c666e9da/x/concentrated-liquidity/fees.go#L255

**Review by p0mvn — APPROVED** [review]
LGTM

File: x/concentrated-liquidity/fees_test.go (hunk 7 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/fees_test.go:775]: Sorry I am not understanding the question

File: x/concentrated-liquidity/fees_test.go (hunk 6 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/fees_test.go:774]: So we currently test for positions that dont have a fee accumulator set. I am not sure if it should be the job of this function to ensure fee accumulator is set at the pool level. Unless I am misunderstanding, that would require us to make a new function to create a cl pool specifically so we dont initialize a fee accumulator.

File: x/concentrated-liquidity/fees_test.go (hunk 7 above)
Comment (mattverse) [reply x/concentrated-liquidity/fees_test.go:773]: AFAIK `updateFeeAccumulatorPosition` can work as either `RemoveFromPositionCustomAcc` or `AddToPositionCustomAcc` depending on whether we use negative / positive value for liquidity Delta, I was curious if we're testing both paths

File: x/concentrated-liquidity/fees_test.go (hunk 6 above)
Comment (mattverse) [reply x/concentrated-liquidity/fees_test.go:774]: This makes sense to me!

**Review by mattverse — APPROVED** [review]
LGTM!

**Pull request merged**